### PR TITLE
check for a setter method 'set[PropertyName]' when property is hydrated

### PR DESCRIPTION
**Pull request by ProxiBlue**
If the code that sets a property checks for existence of a setter method in component class, one can inject logic at the time of setting the property at hydration.

Usage case:

Building a custom pagination component, the current category is added as a queryString to be populated with hydration.

```
class Pagination extends Component\Pagination
{
    public $currentCategory;

    protected $queryString = [
        'page',
        'pageSize',
        'currentCategory'
    ];
```

This works fine, and the currentCategory property is 100% populated on pagination clicks, however, when one tries to load the current product collection, the current layer will default to root category.

```
public function getAllPageItems(): array
    {
        return $this->listProduct->getLoadedProductCollection()->getItems();
    }
```
which end up on 

Magento\Catalog\Block\Product\ListProduct

```
/**
     * Retrieve loaded category collection
     *
     * @return AbstractCollection
     */
    public function getLoadedProductCollection()
    {
        $collection = $this->_getProductCollection();

        $categoryId = $this->getLayer()->getCurrentCategory()->getId();
        foreach ($collection as $product) {
            $product->setData('category_id', $categoryId);
        }

        return $collection;
    }
```

```
$categoryId = $this->getLayer()->getCurrentCategory()->getId();
```

will result to the default root category

so, by using a setter in the Pagination class, which is called instead of placing the property directly, one can add logic to re-populate the base layout category.

in pagination component class:

```
 public function setCurrentCategory($value) {
        $this->currentCategory = (int)$value;
        $this->listProduct->getLayer()->setCurrentCategory((int)$value);
    }
```

The ability to simply intercept property hydration via a setter metod can have other usage case benefits, and not require unneeded plugins/code extending.

This does however place the responsibility to populate the property with the developer making the code. I think that is ok, and would be expected, when placing the setter method. 

image

### CODE REVIEW

**Review by wpoortman — COMMENTED** [review]
Hmm, I'm not sure this should work this way. I get your point, but as far as my knowledge goes, shouldn't this be done with a `updatingCategoryEntity($value)` or `updatedCategoryEntity($value)` method?

Maybe I'm just misunderstanding what you're trying to tell me.

**Comment by ProxiBlue** [comment]
Let me explain in more detail:

When making a pager, that works on category pages:

image

The state for what the current_category is, is naturally lost, due to the fact the pager button click is a new ajax call.
So, to preserve that id (without writing it into session data), it seems logical to pass the current category ID along with the page request:

```
protected $queryString = [
        'page',
        'pageSize',
        'currentCategory',
        'limit'
    ];
```

image

The wire:click goes to its method: toNextpage

image

Now, when the Pagination component is hydrated, and because I can intercept the population of the ```public $currentCategory;```

 public function setCurrentCategory($value)
    {
        $this->currentCategory = (int)$value;
        $this->catalogLayer->setCurrentCategory((int)$value);
        $collection = $this->catalogLayer->getProductCollection();
        $this->getParent()->setCollection($collection);
    }

I can re-setup the collelction to the correct category, else the call to fetch the categoryColelction, will default back to the default ROOT catefory ID

catalogLayer is injected via DI

```
public function __construct(Resolver $layerResolver)
    {
        $this->catalogLayer = $layerResolver->get();
        $this->currentCategory = $this->catalogLayer->getCurrentCategory()->getId();
    }
```

and the eventual run of ToPage will be able to use the correct collection to page on.

```
public function toPage($page): void
    {
        if ($page === $this->getLastPage()) {
            $this->dispatchNoticeMessage(__('You\'ve reached the last page.'));
        }
        if ($page === 1) {
            $this->dispatchNoticeMessage(__('You\'ve reached the first page.'));
        }

        $this->page = (int)$page;
        $collection = $this->catalogLayer->getProductCollection();
        $collection->setCurPage($this->page);
        $this->emitTo('magewire.product.datatable','updateCollection', [$this->page,$this->currentCategory,$this->limit]);
    }
``` 

and I can pass the data to the data table to re-populate the product list as well etc.

Basically, by allowing the ability to intercept the setting of the properties values upon hydration with setters, one gain ability to execute some logic (if wanted) when they are hydrated by a setter. for whatever purpose.

It ads additional ability to do 'something'

**Comment by ProxiBlue** [comment]
as discussed in Slack, this 'feature' could pose a security problem, as anything could be posted into the setter...however, that *could* also be true for just populating the property?

As it stands, this PR just to push for an idea of enhancement, and *should not* be patched by anyone into a production environment, as not all security aspects have been considered

**Comment by ProxiBlue** [comment]
I am closing this PR, as I have realised it will in fact be better to do the actions needed in the final method called when the magewire block in initialised.

 // Finalize component with its Response object
$component->setResponse($this->getHttpFactory()->createResponse($component->getRequest()));

Upon called, all the properties are set to teh values as per teh queryString passed, and one can do the needed logic for anything AFTER those are populated here.
This negates the potential of passed params to the setters manipulated, and poses no security risk.

as an example, I can initialise the toolbar parent block collection to the correct category here

```
/**
     * @param ResponseInterface $response
     * @return $this
     */
    public function setResponse($response): self
    {
        parent::setResponse($response);
        $this->catalogLayer->setCurrentCategory((int)$this->currentCategory);
        $this->initCollection();
        return $this;
    }
```

as everything had been populated.
